### PR TITLE
Remove vote when checkVote() fails in HandleGovernanceVote

### PR DESCRIPTION
**Pull request by ian0371**
## Proposed changes

- Adding an existing validator via `governance.vote("governance.addvalidator", existing_val_addr)` has a bug. Refer to the log below.
- This is caused by https://github.com/klaytn/klaytn/blob/dev/governance/handler.go#L365 where `checkVote()` checks if the address to be added already exists in the validator set, in which case it returns WITHOUT REMOVING THE VOTE.
- This fix removes the vote when `checkVote()` fails. (remove in this context means setting Casted to true)

### Bug log
- Initial setting
For an initial council, we remove one validator so that we can add it back.

```
> klay.getCouncil()
["0x19c032589b76ac395d3a713088e67f9ce9edf1c6", "0x353a5fd7cd9b8d0fe3e7c65d6c3d229cd0f14533", "0xd4ca4f3f2d80fb5b2f569df45d4c28a98f48a289", "0xe0c91bc72396b6c1e937ef717bf59a90c558f285"]
> addr = klay.getCouncil()[0]
"0x19c032589b76ac395d3a713088e67f9ce9edf1c6"
> governance.vote("governance.removevalidator", addr)
"Your vote is prepared. It will be put into the block header or applied when your node generates a block as a proposer."
> governance.myVotes
[{
    BlockNum: 6,
    Casted: true,
    Key: "governance.removevalidator",
    Value: "0x19c032589b76ac395d3a713088e67f9ce9edf1c6"
}]
```

- add validator (1)

```
> governance.vote("governance.addvalidator", addr)
"Your vote is prepared. It will be put into the block header or applied when your node generates a block as a proposer."
> governance.myVotes
[{
    BlockNum: 6,
    Casted: true,
    Key: "governance.removevalidator",
    Value: "0x19c032589b76ac395d3a713088e67f9ce9edf1c6"
}, {
    BlockNum: 16,
    Casted: true,
    Key: "governance.addvalidator",
    Value: "0x19c032589b76ac395d3a713088e67f9ce9edf1c6"
}]
```

- add validator (2) with an existing address

```
> governance.vote("governance.addvalidator", addr)
"Your vote is prepared. It will be put into the block header or applied when your node generates a block as a proposer."
> governance.myVotes
[{
    BlockNum: 0,
    Casted: false,
    Key: "governance.addvalidator",
    Value: "0x19c032589b76ac395d3a713088e67f9ce9edf1c6"
}, {
    BlockNum: 6,
    Casted: true,
    Key: "governance.removevalidator",
    Value: "0x19c032589b76ac395d3a713088e67f9ce9edf1c6"
}]
```

We can see that the second vote is not cast and it keeps appearing in every block the voter becomes the proposer.

```
> klay.getBlock(16).voteData
"0xf84294353a5fd7cd9b8d0fe3e7c65d6c3d229cd0f1453397676f7665726e616e63652e61646476616c696461746f729419c032589b76ac395d3a713088e67f9ce9edf1c6"
> klay.getBlock(22).voteData
"0xf84294353a5fd7cd9b8d0fe3e7c65d6c3d229cd0f1453397676f7665726e616e63652e61646476616c696461746f729419c032589b76ac395d3a713088e67f9ce9edf1c6"
> klay.getBlock(25).voteData
"0xf84294353a5fd7cd9b8d0fe3e7c65d6c3d229cd0f1453397676f7665726e616e63652e61646476616c696461746f729419c032589b76ac395d3a713088e67f9ce9edf1c6"
> klay.getBlock(28).voteData
"0xf84294353a5fd7cd9b8d0fe3e7c65d6c3d229cd0f1453397676f7665726e616e63652e61646476616c696461746f729419c032589b76ac395d3a713088e67f9ce9edf1c6"
```

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by blukat29** [comment]
@ian0371 Can you add a scenario test case in `TestSnapshot_Validators_AddRemove`? (https://github.com/klaytn/klaytn/blob/dev/consensus/istanbul/backend/engine_test.go#L995)
For instance, addval -> addval -> removeval -> removeval case.

I see that unit test for HandleGovernanceVote checks crucial conditions. But having an intuitive scenario as a test would be more comprehensible.

File: governance/handler.go [hunk 1]
```diff
@@ -360,17 +360,33 @@ func (gov *Governance) HandleGovernanceVote(valset istanbul.ValidatorSet, votes
...
+							return false, true
+						}
+					}
+				} else {
+					// invalid value type
+					return true, false
 				}
-			} else {
+				return true, true
+			}
```
Comment (blukat29) [inline governance/handler.go:381]: Can we keep the original if statement? It seems adding `checkValidator` does not add code clarity. It's rather overcomplicated to me.

File: governance/handler.go [hunk 2]
```diff
@@ -360,17 +360,33 @@ func (gov *Governance) HandleGovernanceVote(valset istanbul.ValidatorSet, votes
...
+
+			if checkVotePass, typeAssertPass := checkValidator(); !checkVotePass {
+				// this vote is adding existing validator or removing nonexistent validator
+				if self == proposer {
+					gov.removeDuplicatedVote(gVote, header.Number.Uint64())
+				}
```
Comment (blukat29) [inline governance/handler.go:387]: If you intended to remove duplication, how about defining below function?

```
removeCastedMyVote := func() {
    if self == proposer {
        gov.removeDuplicatedVote(gVote,
    }
}
```

**Comment by hyunsooda** [comment]
Let me ask a simple question. Is it invalid removing duplicated vote immediately from [Vote()](https://github.com/klaytn/klaytn/blob/dev/governance/api.go#L84)? Why did you delegate it to `HandleGovernanceVote()`?

**Comment by hyunsooda** [comment]
Each independent test case seems separated by comment. It looks better to make an arbitrary scope case by case or split them into functions. You may consider this later.

**Comment by ian0371** [comment]
@hyunsooda
> Let me ask a simple question. Is it invalid removing duplicated vote immediately from [Vote()](https://github.com/klaytn/klaytn/blob/dev/governance/api.go#L84)? Why did you delegate it to `HandleGovernanceVote()`?

Good question. Validator set can be fetched from istanbul.snapshot, which is inaccessible from governance package (istanbul/backend package imports governance, so the reverse cannot be done. `HandleGovernanceVote()` receives valset as a parameter for similar reasons).

> Each independent test case seems separated by comment. It looks better to make an arbitrary scope case by case or split them into functions. You may consider this later.

I agree that the current readability is not so great. I'll take this into account later :)

**Comment by hyunsooda** [comment]
Okay. then, does not have a side-effect of this post-processing?

**Comment by ian0371** [comment]
> does not have a side-effect of this post-processing?

No, it does not have a side effect. The removing vote is rarely executed in the real world (i.e., when adding an existing validator).

**Review by blukat29 — APPROVED** [review]

**Comment by hyunsooda** [comment]
Okay.. thank you.

**Review by yoomee1313 — APPROVED** [review]

File: consensus/istanbul/backend/engine_test.go [hunk 3]
```diff
@@ -1115,8 +1161,7 @@ func TestSnapshot_Validators_AddRemove(t *testing.T) {
 				}
 				if v.key == "governance.addvalidator" {
 					for _, i := range indices {
-						nodeKeys = append(nodeKeys, allNodeKeys[i])
-						addrs = append(addrs, allAddrs[i])
+						includeNode(allAddrs[i], allNodeKeys[i])
```
Comment (ehnuje) [inline consensus/istanbul/backend/engine_test.go:1164]: Why are these lines changed?

File: governance/handler.go [hunk 4]
```diff
@@ -362,11 +362,17 @@ func (gov *Governance) HandleGovernanceVote(valset istanbul.ValidatorSet, votes
 			authorize := key == params.AddValidator
 			if addr, ok := gVote.Value.(common.Address); ok {
 				if !gov.checkVote(addr, authorize, valset) {
+					if proposer == self {
+						gov.removeDuplicatedVote(gVote, header.Number.Uint64())
```
Comment (ehnuje) [inline governance/handler.go:366]: How about leaving a log when removing duplicated votes?

File: consensus/istanbul/backend/engine_test.go (hunk 3 above)
Comment (ian0371) [reply consensus/istanbul/backend/engine_test.go:1164]: Without them, the same node ends up twice in `nodeKeys`/`addrs`, resulting in "invalid committed seals".

File: governance/handler.go (hunk 4 above)
Comment (ian0371) [reply governance/handler.go:367]: Added.

**Comment by ian0371** [comment]
Rebased to the latest dev

File: governance/default_test.go [hunk 5]
```diff
@@ -803,6 +803,23 @@ func TestGovernance_HandleGovernanceVote_None_mode(t *testing.T) {
 	}
 	gov.voteMap.Clear()
 
+	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Test adding an existing validator
```
Comment (zacscoding) [inline governance/default_test.go:807]: How about adding remove not exist validator test?

Comment (ian0371) [reply governance/default_test.go:824]: Added. PTAL

File: consensus/istanbul/backend/engine_test.go (hunk 3 above)
Comment (ehnuje) [reply consensus/istanbul/backend/engine_test.go:1194]: @ian0371 What I meant to say was that changing the test code because of the core code change looks strange.
Do we need to sync the test code logic and the core code logic manually for this test?

Comment (ian0371) [reply consensus/istanbul/backend/engine_test.go:1194]: @ehnuje  Changing these lines is inevitable because the existing code cannot handle the new test cases. You can consider it "fixing a bug". 
I could add a new test function to verify them, but instead, I decided to reuse the existing function because most parts (i.e., except for these lines) will be reused. You might not want to go down that road.

**Review by blukat29 — APPROVED** [review]

**Review by zacscoding — APPROVED** [review]
Thanks @ian0371  👍
